nifi-nar-bundles/nifi-social-media-bundle/nifi-twitter-processors/src/main/java/org/apache/nifi/processors/twitter/ConsumeTwitter.java

```diff
@@ -0,0 +1,316 @@
+            .required(true)
+            .defaultValue("0")
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("All tweets will be routed to this relationship.")
+            .build();
+
+    private List<PropertyDescriptor> descriptors;
+    private Set<Relationship> relationships;
+
+    private TwitterStreamAPI api;
+
+    private volatile BlockingQueue<String> messageQueue;
+
+    private void emptyQueue() {
+        while (!messageQueue.isEmpty()) {
+            messageQueue.poll();
+        }
+    }
+
+    @Override
+    protected void init(ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(ENDPOINT);
+        descriptors.add(BASE_PATH);
+        descriptors.add(BEARER_TOKEN);
+        descriptors.add(QUEUE_SIZE);
+        descriptors.add(BATCH_SIZE);
+        descriptors.add(TWEET_FIELDS);
+        descriptors.add(USER_FIELDS);
+        descriptors.add(MEDIA_FIELDS);
+        descriptors.add(POLL_FIELDS);
+        descriptors.add(PLACE_FIELDS);
+        descriptors.add(EXPANSIONS);
+        descriptors.add(BACKFILL_MINUTES);
+
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return this.descriptors;
+    }
+
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        messageQueue = new LinkedBlockingQueue<>(context.getProperty(QUEUE_SIZE).asInteger());
+
+        api = new TwitterStreamAPI(context, messageQueue, getLogger());
+        api.setBasePath(context.getProperty(BASE_PATH).getValue());
+        final String endpointName = context.getProperty(ENDPOINT).getValue();
+        if (ENDPOINT_SAMPLE.getValue().equals(endpointName)) {
+            api.start(TwitterStreamAPI.SAMPLE_ENDPOINT);
+        } else if (ENDPOINT_SEARCH.getValue().equals(endpointName)) {
+            api.start(TwitterStreamAPI.SEARCH_ENDPOINT);
+        } else {
+            throw new AssertionError("Endpoint was invalid value: " + endpointName);
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        if (messageQueue.isEmpty()) {
+            context.yield();
+            return;
+        }
+
+        FlowFile flowFile = session.create();
+        flowFile = session.write(flowFile, new OutputStreamCallback() {
```

Comment:
It would be safer to use Jackson to convert the incoming tweets into `Map<String, Object>` types and then turn the current queue contents into a stringified array with Jackson.

Reply:
Thanks for the comment. A few thoughts. Would that consume too much memory though? Especially if the batch size is small and the maximum queue size is large, the `BlockingQueue` that the thread uses to store the tweets could be larger as opposed to when the contents of the queue were just a string.

Also, as of current writing, I'm sure enough that we don't have to worry about malformed tweets as part of the stream that is returned by the API. If the GET request for the stream is invalid, the underlying API provided by the Twitter Java SDK throws an `ApiException`. Since that's the case writing strings into the `OutputStream` directly seems efficient.

Reply:
Fair enough.

Reply:
Concurring with @emiliosetiadarma, writing the strings directly from the Twitter API response stream seems like the optimal approach. The Twitter SDK follows the same strategy, unfortunately it doesn't provide a clean way to control the lifecycle of the thread responsible for reading the stream. Although this approach requires expecting each string to contain a well-formed object, this seems acceptable given the performance considerations.